### DISCUSSION TITLE
Best approach for email verification in credentials

### DISCUSSION THREAD

**Discussion by arashi-dev** [nextauthjs/next-auth#3478] [Help] [answered]
Hello. 
I'm newbie to `NextAuth.js`. I am creating custom credentials for `sign up` `login in` `reset password` and etc.

**Question**
my question is how can I handle email verification? 

**My thoughts**
I think I should send email via `nodemailer` and create a link which has a `JWT` token in it and valid for 24/48 hours and redirects user to a nextjs page. 
and in that page, I have the nextauth `signIn` function which calls the `email-verification` credential and sends the JWT token to the nextauth.
then, in the credential Provider, I verify the token and get the user id that was stored in the token and verify the email in database.

**So**
Is it a correct solution?

**Comment by balazsorban44** [comment]
https://next-auth.js.org/providers/email

**Reply by arashi-dev** [reply]
Thanks for your response.
But can I customize it? I already have an user model in database. and I think Email Provider creates its own model. not updating mine.

**Comment by alegatt** [comment]
Hi! Have you found a solution to this problem?

**Reply by arashi-dev** [reply]
yes I did. sorry I forgot to answer my question. 
https://github.com/nextauthjs/next-auth/discussions/3478#discussioncomment-1935169

**Comment by choutkamartin** [comment, marked as answer]
Assuming you really mean credentials provider and not an email one:
Create a sign-up page, allow the user to enter an email, send a request to your API, check if the email exists in your database, and if not, then create a random hash and store that hash along with user-entered email in the database. Then do a check if the save was successful. If it was, send an email to the address user has entered. The link should contain a path to your page that verifies the token and the token as a query string. The page will call your API with fetch inside `useEffect` or `getServerSideProps`. The API will then check if the verification token sent in the API request from the query string matches the one stored in the database. If they match, switch the `isVerified` property to true. (Or use `email_verified` property Next Auth uses.) You will use this `isVerified` property inside `authorize()` method and either allow the user to sign in, or you will show "Not verified" error message.

**Reply by arashi-dev** [reply]
yes, I did almost what you said.

I made a credential and named it "verify-email" which gets a JWT token.
then, I call signIn("verify-email") and pass the token to it. 
now, in verify-email credential, the token will be verified and updates the `emailVerifiedAt` field user in database.

and the JWT token contains the user id as payload and a `string + email` as a secret which makes the token unpredictable for attackers and one-time for usage.

**Reply by hayrullahuseynovbusiness** [reply]
hi, i have the same problem could you show your code

**Reply by arashi-dev** [reply]
@hayrullahuseynovbusiness 

when user signs up with credential provider, an email will be sent to user which contains a link like so:
```ts
"http://localhost/account/verify-email?token=abc123"
```
and this is how I generated token:
```ts
const token = jwt.sign({ id }, process.env.NEXTAUTH_PROVIDER_EMAIL_VERIFICATION_SECRET + email, {
    expiresIn: "1d"
})
```

and then, I created another credential provider which accepts `token: string`.
It decodes and verifies the token.
```ts
const isValid = await new Promise((resolve) => {
    jwt.verify(token, process.env.NEXTAUTH_PROVIDER_EMAIL_VERIFICATION_SECRET + email, (err) => {
        if (err) resolve(false)
        if (!err) resolve(true)
    })
})
```
in this way, if user has changed its email address, the old token will be unusable.
so, if the email was changed or the token was expired, it will throw an error and the `isValid` will be equals to false. otherwise, it will be true.

finally, I will check if the `isValid` was true, I will update user data in database like so `emailVerifiedAt: new Date()`. and login the user by returning an object which contains some user info:
```ts
return { id }
```

**Reply by HynekS** [reply]
There's a [pretty neat video made by Ben Awad in 2017](https://www.youtube.com/watch?v=76tKpVbjhu8) where he is dealing with email verification (virtually the same way as described above).

**Comment by riolly** [comment]
I am looking for this answer until I realize it.
Why OAuth exist is to eliminate this process. We can log in securely without verification.
If you use OAuth then you have to open the email and verify then what is the point?

The case that I think need verification is for example when a user is invited for a certain role. Then we can send the invitation email with the verification link to assign the role.

Am I wrong?

**Reply by arashi-dev** [reply]
sort of.
not everyone has Gmail or Facebook account. Additionally, not everyone interests to use Gmail for the website. maybe they don't want to use their primary account or something. this happened to me already many times.

but anyway, I really like your idea. I mean, I hope someday soon the email approach becomes deprecated and non-standard. that is nothing but a loss of resources and times

**Reply by kasparlaurenz** [reply]
Hey would you mind creating a public repo with your approach. Id like to implement email verification aswell.

**Reply by arashi-dev** [reply]
I don't have the source code, unfortunately. I deleted the code because I changed the backend to `Strapi`. I don't have that. 
But I can help if you need me.

**Reply by kasparlaurenz** [reply]
Ah i see. I appreciate your offer :) do you have a discord or any?

**Reply by arashi-dev** [reply]
yes. arashi-dev#7753 is my discord id

**Reply by nimeshmaharjan1** [reply]
> I am looking for this answer until I realize it. Why OAuth exist is to eliminate this process. We can log in securely without verification. If you use OAuth then you have to open the email and verify then what is the point?
> 
> The case that I think need verification is for example when a user is invited for a certain role. Then we can send the invitation email with the verification link to assign the role.
> 
> Am I wrong?

But what about authentication for the admin where you need the username and password you cant just make the admin login with the oauth or send a email everytime he/she logs in right?
